core/internal/user/aggregate.go

```diff
@@ -0,0 +1,119 @@
+package user
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+
+	"github.com/get-eventually/go-eventually/core/aggregate"
+	"github.com/get-eventually/go-eventually/core/event"
+)
+
+type WasCreated struct {
+	ID        uuid.UUID
+	FirstName string
+	LastName  string
+	BirthDate time.Time
+	Email     string
+}
+
+func (WasCreated) Name() string {
+	return "UserWasCreated"
+}
+
+type EmailWasUpdated struct {
+	Email string
+}
+
+func (EmailWasUpdated) Name() string {
+	return "UserEmailWasUpdated"
+}
+
+var _ aggregate.Root[uuid.UUID] = &User{}
+
+type User struct {
+	aggregate.BaseRoot
+
+	// Aggregate field should remain unexported if possible,
+	// to enforce encapsulation.
+
+	id        uuid.UUID
+	firstName string
+	lastName  string
+	birthDate time.Time
+	email     string
+}
+
+func (user *User) Apply(event event.Event) error {
+	switch evt := event.(type) {
+	case WasCreated:
+		user.id = evt.ID
+		user.firstName = evt.FirstName
+		user.lastName = evt.LastName
+		user.birthDate = evt.BirthDate
+		user.email = evt.Email
+
+	case EmailWasUpdated:
+		user.email = evt.Email
+
+	default:
+		return fmt.Errorf("%T: unexpected event type, %T", user, event)
+	}
+
+	return nil
+}
+
+func (user *User) AggregateID() uuid.UUID {
+	return user.id
+}
+
+func Create(id uuid.UUID, firstName, lastName, email string, birthDate time.Time) (*User, error) {
+	user := &User{}
+
+	if firstName == "" {
+		return nil, fmt.Errorf("%T: invalid first name, is empty", user)
+	}
+
+	if lastName == "" {
+		return nil, fmt.Errorf("%T: invalid last name, is empty", user)
+	}
+
+	if email == "" {
+		return nil, fmt.Errorf("%T: invalid email, is empty", user)
+	}
+
+	if birthDate.IsZero() {
+		return nil, fmt.Errorf("%T: invalid birth date, not specified", user)
+	}
+
+	if err := aggregate.RecordThat[uuid.UUID](user, event.Envelope{
+		Message: WasCreated{
+			ID:        id,
+			FirstName: firstName,
+			LastName:  lastName,
+			BirthDate: birthDate,
+			Email:     email,
+		},
+	}); err != nil {
+		return nil, fmt.Errorf("%T: failed to record domain event, %w", user, err)
+	}
+
+	return user, nil
+}
+
+func (user *User) UpdateEmail(email string) error {
+	if email == "" {
+		return fmt.Errorf("%T: invalid email, is empty", user)
+	}
+
+	if err := aggregate.RecordThat[uuid.UUID](user, event.Envelope{
```

Comment:
Yeah, although I'm sure this would probably happen in other languages too.

The reason for it is: `aggregate.Root` has one generic parameter for its ID (`aggregate.ID` constraint), to give correctly-typed return value when calling `root.AggregateID()`.

However, `aggregate.RecordThat`, which accepts a generic `aggregate.Root[I]`, needs `I` to be explicitly specified, so that it can validate that the generic `T` parameter used actually implements `aggregate.Root[I]`.

---

EDIT: now that I think about it, Rust wouldn't have this issue since one could use [associated types](https://doc.rust-lang.org/rust-by-example/generics/assoc_items/types.html) over an `AggregateRoot` trait to specify the `Id`. So the type-system can always resolve the associated `AggregateRoot::Id` type from a concrete type implementing the trait.

[(Well, this is exactly how it is implemented in the Rust version of the lib)](https://github.com/get-eventually/eventually-rs/blob/5374c0aeb550ab97434cce9c67f6b33e4ccfd6e8/eventually/src/aggregate.rs#L51-L53)

I think however the Java/Kotlin type-system is missing this, and probably has the same limitation Go has 👀

Reply:
From a compiler perspective, I'm pretty sure even vanilla Java would infer this type correctly. In this particular case, one could even use wildcards as the type is not really referenced anywhere, i.e.
```java
public void RecordThat(Root<? extends ID> root, Event... events) { }
```
Associated types would be a better choice if it was possible, yeah.
In any case, I still don't see how anything is really checked here tbh, considering that `Aggregate` is a non-generic interface and `root` is always type-compatible with itself. If it's supposed to be a responsibility of a user to make sure a concrete Command Handler implementation passes the matching ID type, then it looks like generics aren't really helping. Or I don't understand the code.